### ISSUE TITLE
Support `scoped` modifier for other local declarations

### ISSUE THREAD

**Issue by cston** [dotnet/roslyn#62039]
This issue was discussed in LDM 7/27/2022. It is blocked on https://github.com/dotnet/roslyn/issues/62983 (spec existing grammar and consider impact on implementation).

```csharp
// foreach:
foreach (scoped R r in c1) { }
foreach (scoped ref int i in c2) { }

// for:
for (scoped R r = s; ; ) { }
for (scoped ref int a = ref b; ; ) { }

// deconstruction:
(scoped R x, scoped R y) = F();
scoped var (x, y) = F(); // probably should disallow

// out var:
F(out scoped R r);
F(out scoped var r);
```

Relates to test plan https://github.com/dotnet/roslyn/issues/59194

To support this, we'd need to expand the grammar to allow modifiers in `declaration_expression`, `for_each_statement`, `for_statement` (see proposal below).

This raises the questions:
1. do we want to do this?
2. confirm that `scoped` would be the only modifier allowed in those positions. Other declaration modifiers (`const`, `static`, `readonly`, `volatile`) are already not applicable to local declarations.

```antlr
declaration_expression
  : modifier* type variable_designation // add modifiers
  ;

for_each_statement
  : attribute_list* 'await'? 'foreach' '(' modifier* type identifier_token 'in' expression ')' statement // add modifiers
  ;

for_each_variable_statement
  : attribute_list* 'await'? 'foreach' '(' expression 'in' expression ')' statement // covered by declaration_expression
  ;


fixed_statement
  : attribute_list* 'fixed' '(' variable_declaration ')' statement // probably leave untouched
  ;

for_statement
  : attribute_list* 'for' '(' (modifier* variable_declaration? | (expression (',' expression)*)?) ';' expression? ';' (expression (',' expression)*)? ')' statement // add modifiers
  ;
  
// Local declarations and parameters already support modifiers:

local_declaration_statement
  : attribute_list* 'await'? 'using'? modifier* variable_declaration ';'
  ;
```

**Comment by jcouv** [comment]
Re-assigned to @cston who is driving discussion on grammar and syntax model. Those design questions need to be resolved before we can make this change.
Otherwise, LDM is supportive of this, but details will need to be figured out.

**Comment by 333fred** [comment]
Discussed in LDM: https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-07-27.md

**Comment by cston** [comment]
See https://github.com/dotnet/csharplang/pull/6338 for proposed grammar.
See https://github.com/dotnet/roslyn/issues/61647 for `ScopedTypeSyntax` for locals.

**Comment by RikkiGibson** [comment]
I have associated this issue with the test plan item: *check internal enumerations of local variable kinds*. It seems like the process of checking local declarations which can be marked scoped goes hand in hand with reviewing the internal enumerations.
